dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
         String.format("Rule '%s' does not exist. Create it first with a POST request.", id)
       );
     }
 
-    return String.format(RULE_TEMPLATE, row.getString(ID), row.getString(DESCRIPTION),
-                                row.getString(CONDITION), row.getString(ACTION));
+    StructuredRow row = optionalStructuredRow.get();
+
+    return String.format(RULE_TEMPLATE, row.getString(ID_COL), row.getString(DESCRIPTION_COL),
+                         row.getString(CONDITION_COL), row.getString(ACTION_COL));
   }
 
-  public void createRulebook(RulebookRequest rb) throws RulebookAlreadyExistsException, RuleNotFoundException {
-    Row row = rulebook.get(toKey(rb.getId()));
-    if (!row.isEmpty()) {
+  void createRulebook(RulebookRequest rulebookRequest)
+    throws RulebookAlreadyExistsException, RuleNotFoundException, IOException {
+    Collection<Field<?>> rulebookKeyFields = new ArrayList<>();
+    rulebookKeyFields.add(Fields.stringField(ID_COL, rulebookRequest.getId()));
+
+    Optional<StructuredRow> optionalRulebookStructuredRow = rulebookTable.read(rulebookKeyFields);
+
+    if (optionalRulebookStructuredRow.isPresent()) {
       throw new RulebookAlreadyExistsException(
-        String.format("Rulebook '%s' already exists. Either update it or delete it and create it.", rb.getId())
+        String
+          .format("Rulebook '%s' already exists. Either update it or delete it and create it.", rulebookRequest.getId())
       );
     }
 
-    for (String rule : rb.getRules()) {
-      if(rules.get(toKey(rule)).isEmpty()) {
+    for (String rule : rulebookRequest.getRules()) {
+      Collection<Field<?>> ruleKeyFields = new ArrayList<>();
+      ruleKeyFields.add(Fields.stringField(ID_COL, rule));
+
+      Optional<StructuredRow> optionalRuleStructuredRow = rulesTable.read(ruleKeyFields);
+
+      if (!optionalRuleStructuredRow.isPresent()) {
         throw new RuleNotFoundException(
           String.format("Rulebook '%s' includes a rule '%s' that does not exist. Please add rule first",
-                        rb.getId(), rule)
+                        rulebookRequest.getId(), rule)
         );
       }
     }
 
-    byte[][] columns = new byte[][] {
-      ID, DESCRIPTION, CREATED, UPDATED, USER, SOURCE, VERSION, RULES
-    };
-
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(rb.getId()),
-      Bytes.toBytes(rb.getDescription()),
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(rb.getUser()),
-      Bytes.toBytes(rb.getSource()),
-      Bytes.toBytes(1L),
-      Bytes.toBytes(rb.getRulesString())
-    };
-
-    rulebook.put(toKey(rb.getId()), columns, values);
+    Collection<Field<?>> rulebookFields = new ArrayList<>();
+    rulebookFields.add(Fields.stringField(ID_COL, rulebookRequest.getId()));
+    rulebookFields.add(Fields.stringField(DESCRIPTION_COL, rulebookRequest.getDescription()));
+    rulebookFields.add(Fields.longField(CREATED_COL, getCurrentTime()));
```

Comment:
similar comment about using the same value for created and updated

Reply:
Done